### DISCUSSION TITLE
fuzzy search and results ranked by relevance

### DISCUSSION THREAD

**Discussion by schmity** [ldproxy/ldproxy#930] [Ideas]
We are currently working on optimizing search capabilities within our published datasets, i.e. object-level. Starting with search for streets and addresses, we plan to extend this capability to other datasets in the future. It is a strong requirement by our users to have some kind of fuzzy search (e.g. typos) and to get the results ranked by relevance. 
At the same time we would like to publish such a search API via OGC API - Features, e.g. via the ?q= parameter available in ldproxy.

Our data resides in Postgres and we have looked at the different search capabilities Postgres offers. For our current use case, we got promising results using [pg_trgm](https://www.postgresql.org/docs/13/pgtrgm.html) while [Postgres Full Text Search (FTS)](https://www.postgresql.org/docs/13/textsearch.html) may be useful for others.
Would it be within the scope of ldproxy to adapt the query generation to these Postgres features as additional configuration options for the SQL provider, e.g. using pg_trgm operators (%, <->) rather than LIKE or ILIKE? 
An alternative would be to use Elastic Search, even though that would introduce a new component to operate. Would it be within the scope of ldproxy to add Elastic search as provider or to proxy requests to it similar to the WFS provider?

Best,
Sebastian

##
Urban Data Platform Hamburg

**Comment by cportele** [comment]
@schmity 

The first step would be to figure out the API design for the requirements, independent of the implementation. 

The `q` parameter as it is in the drafts does not address your requirements, since `q` only selects objects, but does not rank them. `q` is currently "just" a filter parameter and sorting has to be specified using the `sortby` parameter. The server is of course free to return the query results in any order - unless `sortby` has been specified. This may also be a topic to discuss in OGC, but for now let's assume that `q` and `sortby` will not change.

Applying fuzzy search and returning results ranked by relevance would then be at least a separate parameter that could not be used with other filter or sort parameters on the `items` endpoint. A separate resource/endpoint would probably be a better approach.

The second step could then be to discuss how this could be implemented in PostgreSQL. It is certainly an option to define CQL2 functions that map to, e.g., operators of pg_trgm. Adding an Elastic Search provider would be an option, too. It would require an analysis how CQL2 maps to the native capabilities - including the spatial support, but in general that should be possible (AFAIK, some STAC CQL2 implementations use Elastic Search as backends; for example, the requirement in CQL2 Basic to always have a property on the left side of an operator and no property on the right side was added because of the way queries work in Elastic Search).

Would it be best to discuss this is a call?

Best regards, Clemens
